### ISSUE TITLE
Does not find members when casting handle in with-statement

### ISSUE THREAD

**Issue by postlund** [PyCQA/pylint#4534]
### Steps to reproduce

Given a file `a.py`:
```python
import wave
from typing import cast

with wave.open("dummy.wav", "wb") as handle:
    writer = cast(wave.Wave_write, handle)
    writer.setnchannels(2)
    writer.setsampwidth(2)
    writer.setframerate(44100)
```

### Current behavior

Result of `pylint a.py`:
```
a.py:7:4: E1101: Instance of 'Wave_read' has no 'setnchannels' member; maybe 'getnchannels'? (no-member)
a.py:8:4: E1101: Instance of 'Wave_read' has no 'setsampwidth' member; maybe 'getsampwidth'? (no-member)
a.py:9:4: E1101: Instance of 'Wave_read' has no 'setframerate' member; maybe 'getframerate'? (no-member)
```

### Expected behavior

No warnings expected. `mypy` does not give any errors.

### pylint --version output

Result of `pylint --version` output:

```
pylint 3.0.0a3
astroid 2.5.7
Python 3.8.1 (default, Jan  8 2020, 13:13:59) 
[GCC 4.9.2]
```

**Comment by hippo91** [comment]
@postlund thanks for the report. Could you try without the call to cast? It would determine if the problem is linked to the inference of cast function or to the wave module.

**Comment by postlund** [comment]
@hippo91 I get the same error if I remove the cast by the looks of it.

**Comment by timmartin** [comment]
It looks to me like there's no logic currently to infer the result of the `cast()` function. IIUC this is something that should be solved in Astroid?

I had a go at implementing this [here](https://github.com/timmartin/astroid/commit/3e7697f6e05683bdd9c6e3f3c376ffa573a3d65a), I think this will need a bit of work to make it more rigorous but it seems like it works in a simple case.

**Comment by Pierre-Sassoulas** [comment]
@timmartin, could you open a PR in astroid please ? Apparently bad inference in cast is not at fault for this issue, but this is still a nice feature to have :)

**Comment by timmartin** [comment]
I will open a PR, I just wanted to do a little more review of the code first, it's probably not ready to merge.

If I'm understanding this correctly, lack of inference on `typing.cast` is causing this bug. It looks like if the cast was inferred correctly, Pylint would know to treat the `writer` as an instance of `wave.Wave_write`, which would presumably then have the expected methods.

**Comment by jacobtylerwalls** [comment]
See discussion at https://github.com/PyCQA/astroid/pull/1076#issuecomment-883583488 for subtle issues with `TypeVar` that a proposal would need to address.
